### ISSUE TITLE
On firefox, identifying third-party resources only after reload

### ISSUE THREAD

**Issue by gmacon** [EFForg/privacybadger#1101]
When I visit a URL for the first time, Privacy Badger reports no trackers blocked, and the detail pane reports "There are no third-party trackers on this site." Here's an example:

![screen shot 2016-12-20 at 11 54 58 am](https://cloud.githubusercontent.com/assets/238853/21359946/efa9efbc-c6ab-11e6-9729-a124af047ac0.png)

After reloading the page, I get this instead:

![screen shot 2016-12-20 at 11 55 34 am](https://cloud.githubusercontent.com/assets/238853/21359965/0075b72c-c6ac-11e6-9f4d-dacf1ad584b5.png)

Firefox 50.1.0 on macOS
Privacy Badger 2016.12.15.1
Other extensions: HTTPS Everywhere, LastPass, Reddit Enhancement Suite, VimFx

**Comment by iangreenleaf** [comment]
Exactly my issue! Has made browsing the web really unbearable for the past couple weeks.

Firefox 50.1.0 on Arch Linux
Privacy Badger 2016.12.15.1
Other extensions: HTTPS Everywhere, Lastpass, Greasemonkey, Searchbar Autosizer, Tab Mix Plus

**Comment by ungutknut** [comment]
Same problem here since about a month. It's a huge annoyance.

Firefox 50.1.0 on Xubuntu 16.04 64bit
PB 2016.12.15.1
No further extensions

**Comment by gfab23** [comment]
As far as I see it, it does load 3rd party sites, which had been blocked in the past. I think this behavior started around with the update to Firefox 50.1.0. I am running on OS X 10.11.

When I visit sites out of my history or bookmarks PB works properly, e.g. to https://www.heise.de/newsticker/. Also when following any of the links to articles it works. But when I click on a not before visited URL on that site (e.g. https://www.heise.de/security/meldung/33C3-Hartnaeckige-Datenschutzprobleme-bei-Einlogg-Dienst-per-Facebook-Co-3582726.html?wt_mc=rss.security.beitrag.atom) I got out of the e-mail from my RSS feed reader, then PB does not properly work and does not block tracking 3rd party sites. Even Certificate Patrol is reporting on changed certificates for 3rd party sites I have not visited (loaded) for a longer time. See this screen shot from the first visit to above RSS URL:
![eff_privacybadger_first-visit-20161229](https://cloud.githubusercontent.com/assets/17186739/21545428/f40fd21e-cdd7-11e6-9a71-5cf090131f58.png)

Only after a reload PB is working (the advertisements are gone):
![eff_privacybadger_reload-20161229](https://cloud.githubusercontent.com/assets/17186739/21545454/3169ccfa-cdd8-11e6-8b64-1108c1a2db67.png)

**Comment by brbrewster** [comment]
I've been experiencing basically the same problem, but it seems to happen when I open a page in a new tab. The page is extremely slow to load, and trackers aren't detected or blocked. After a page refresh, trackers and ads are both detected and blocked.

This is on Ubuntu 14.04 with Firefox 50.1.0.

**Comment by cooperq** [comment]
Okay this is extermely strange indeed. I am able to reproduce this in 50.1.0 myself. Suspect it's a bug in web extensions will have to do some digging to figure out where. @ghostwords any ideas?

**Comment by ghostwords** [comment]
Looks like `onBeforeRequest` never fires for frame 0 for pages I control-click to open into a background tab. Since `tabData` never [gets populated](https://github.com/EFForg/privacybadger/blob/6f0745bb9af0d3b4fa78a7217e8c17b3ccf32763/src/webrequest.js#L58-L60) for frame 0, this breaks processing potential trackers ([`checkAction` always returns false](https://github.com/EFForg/privacybadger/blob/6f0745bb9af0d3b4fa78a7217e8c17b3ccf32763/src/webrequest.js#L486)), and I think also retrieving data for the browser popup.

**Comment by ghostwords** [comment]
If your findings agree with mine, I suggest making a minimal addon to demonstrate the problem to Firefox Web Extensions people. Do other Web Extensions `chrome.webRequest`-using addons also run into this?

**Comment by ghostwords** [comment]
Is [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1293583) the relevant Firefox bug?

**Comment by ghostwords** [comment]
Oh sorry, looks like I was looking into issue #1078, which may or may not be the same as this one.

**Comment by cooperq** [comment]
I think it's probably the same issue. I bet onBeforeRequest isn't getting fired, that's definitely what it looks like from the outside.

**Comment by ghostwords** [comment]
@groovecoder Could you take a look please?

**Comment by ghostwords** [comment]
I filed a new [Bugzilla issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1329299).

**Comment by groovecoder** [comment]
I'm unable to reproduce this on Mac OS Firefox 51.0b12 (beta) running a fresh profile with no other extensions.

1. When I load a page, trackers are blocked the first time a page is loaded
2. When I ctrl+click to open a link in a new tab, trackers are blocked the first time a page is loaded
3. When I add a `console.log` statement to the `onBeforeRequest` listener in both `webrequest.js` and `background.js`, I see many statements in both of the above cases.

I briefly saw an error `tabId is undefined` at `background.js:429:5` (`return Object.keys(this.tabData[tabId].trackers);`)

I'll keep digging to try to reproduce.

**Comment by mixedpuppy** [comment]
I've commented on the bugzilla bug requesting some terminology clarification and further verification that this only affects 50.

**Comment by ghostwords** [comment]
I couldn't reproduce with a fresh profile in 51.0b12 (beta) and 52.0a2.en-US.linux-x86_64 (dev). Then I tried a fresh profile with Firefox 50.1, and Privacy Badger seemed fine there too, pages opened into new tabs and pages opened directly. Can still reproduce with my existing browser profile with Firefox 50.1 ... I wonder what the difference is.

**Comment by cooperq** [comment]
Cool, looks like we need to do some more testing then.

**Comment by ghostwords** [comment]
Now I can reproduce in a fresh profile with Firefox 51 on all pages, opened into a new tab or not. Same laptop, same Firefox as three hours ago ... no idea what governs whether I get the bug or not.

To recap, the problem is that while we do get a details object of type "main_frame" as expected, its frameId is not 0, which breaks expectations. For example:

![screenshot from 2017-01-09 19 03 32](https://cloud.githubusercontent.com/assets/794578/21788898/50693fe4-d69f-11e6-8ef9-94e9162fba43.png)

**Comment by mixedpuppy** [comment]
@ghostwords is that consistent?  ie. you **always** get a call to the listener as expected, but the frame_id is wrong, and you were filtering on frame_id == 0?

**Comment by ghostwords** [comment]
Yes, I think so. We should never get a details object of type "main_frame" and a non-zero frameId, no?

**Comment by mixedpuppy** [comment]
Is my last response in the bugzilla bug acceptable?

**Comment by mixedpuppy** [comment]
ie. due to timing and inability to reproduce on 52/53, I'm inclined to wontfix the problem, but you'd have to address that on your end.

**Comment by cooperq** [comment]
@mixedpuppy  Do you think it would be possible to track down the patch that fixes this in 52 and uplift it to 51? I can post this comment in bugzilla if that is more appropriate.

**Comment by mixedpuppy** [comment]
My assumption is that you can fix this on your end by not relying on frameId in combination with main_frame, is that not correct?  Uplift to 51 would be a huge quagmire at this point.

**Comment by ghostwords** [comment]
@mixedpuppy Yep, that sounds good. I'll open a Privacy Badger pull request shortly.

**Comment by cooperq** [comment]
fixed by #1131
